./github/scripts/auto-backport.py: don't remove backport label when backport process has an error

yaronkaikov opened the pull request:
Today, when the `Fixes` prefix is missing or the developer is not a collaborator with `scylladbbot` we remove the backport labels to prevent the process from starting and notifying the developers.

Developers are worried that removing these backport labels will cause us to forget we need to do these backports. @nyh suggested to add a `scylladbbot/backport_error` label instead

Applied those changes, so when a `Fixes` prefix is missing we will add a `scylladbbot/backport_error` label and stop the process

When a user doesn't accept the invite we will still open the PR but he will not be assigned and will not be able to edit the branch when we have conflicts

Fixes: https://github.com/scylladb/scylla-pkg/issues/4898
Fixes: https://github.com/scylladb/scylla-pkg/issues/4897

**Backport automation improvements, no need for backport**

tchaikov commented on .github/workflows/add-label-when-promoted.yaml:
```diff
@@ -64,8 +64,17 @@ jobs:
+        id: check_backport_error_label
+        run: |
+          label_name="${{ github.event.label.name }}"
+          if [[ "$label_name" == scylladbbot/backport_error ]]; then
```
just wanted to be more explicit, if we really want to check if this label was *removed*, we should probably instead put
```yaml
run: |
  if [[ "${{ github.event_name }} == 'pull_request_target' ]] && \
     [[ "${{ github.event.action }} == 'unlabeled' ]] &&
     [[ "${{ github.event.label.name == 'scylladbbot/backport_error ]]; then
    echo "backport_error_label=true" >> $GITHUB_OUTPUT
  else
    echo "backport_error_label=false" >> $GITHUB_OUTPUT
  fi
```
also, would be better to have a more readable variable name for `backport_error_label`. i'd suggest use something like "ready_for_backport".

yaronkaikov replied:
send a v3 to simplify it

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +185,24 @@ def main():
-            create_pr_comment_and_remove_label(pr, comment)
-            continue
+            comment += f'Please check your inbox and approve the invitation, otherwise you will not be able to edit PR branch when needed\n'
+            pr.add_to_labels("scylladbbot/backport_error")
```
better off adding a comment explaining the protocol: 

When a pull request is pending for backport but its author is not yet a collaborator of "scylladbbot", we attach a "scylladbbot/backport_error" label to the PR. This prevents the workflow from proceeding with the backport process until the author has been granted proper permissions -- the author should remove the label manually to re-trigger the backport workflow.

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +185,24 @@ def main():
-            version = backport_label.replace('backport/', '')
-            backport_base_branch = backport_label.replace('backport/', backport_branch)
-            backport(repo, pr, version, commits, backport_base_branch)
+        if 'scylladbbot/backport_error' not in labels:
```
this should not happen, so i'd suggest error out if this label is included in `labels`.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci/169/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/releng-testing/job/gating-dtest-release-with-tablets/78/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/releng-testing/job/gating-dtest-release-with-consistent-topology-changes/77/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/releng-testing/job/scylla-ci/169/)
#### Build Details:
- Duration: 7 hr 21 min
- Builder: spider2.cloudius-systems.com

yaronkaikov commented:
v2:
* implement PR comments
* Modify workflow to check if `backport/<release>` is available through all labels

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +185,29 @@ def main():
             print(f'no backport label: {pr.number}')
             continue
-        if args.commits and not with_github_keyword_prefix(repo, pr):
+        if not with_github_keyword_prefix(repo, pr):
```
nit, i'd suggest adding the comment and label here. so the code is more symmetric, like:
```py
backport_labels = [label for label in labels if backport_label_pattern.match(label)]
# ...
if not with_github_keyword_prefix(repo, pr):
  pr.add_to_labels("scylladbbot/backport_error")
  pr.create_issue_comment("...")
  # Fixes tag missing in body, cannot create the backport for this PR
  continue
is_collaborator = repo.private or scylladbbot_repo.has_in_collaborators(pr.user.login)
if not is_collaborator:
  scylladbbot_repo.add_to_collaborators(pr.user.login)
  pr.add_to_labels("scylladbbot/backport_error")
  pr.create_issue_comment("...")
  # author is not a collaborator yet, stop here as well
  continue

# this PR is being promoted, and should be ready to be backported. 
for backport_label in backport_labels:
  # @yaronkaikov if the author is not a collaborator, the current PR should be attached with "scylladbbot/backport_error" label in the above code. but we check
  # for this label and skip a PR if it has this label. why do we still pass and check `is_collaborator` parameter to `backport()` ? it should be always `True`, i think.
  backport(repo, pr, version, commits, backport_base_branch, is_collaborator)
```

yaronkaikov commented:
v4:
* updated strings to multiline where needed (based on PR comments)
* Add `--github-event` parameter to identify when we have `unlabeled` event
* moved comment outside `with_github_keyword_prefix` as suggested

@tchaikov all comments were address

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +180,34 @@ def main():
-        if args.commits and not with_github_keyword_prefix(repo, pr):
+        if not with_github_keyword_prefix(repo, pr) and 'unlabeled' not in args.github_event:
+            comment = f''':warning:  @{pr.user.login} PR body or PR commits do not contain a Fixes reference to an issue and can not be backported
+            please update commit message or PR body with a valid ref to an issue. Then remove `scylladbbot/backport_error` label to re-trigger the backport process
```
since the change has been promoted, force-push to an already-merged PR could be confusing. can we remove "commit message or" from the message?

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +180,34 @@ def main():
+            pr.create_issue_comment(comment)
+            is_collaborator = False
         commits = get_pr_commits(repo, pr, stable_branch, start_commit)
+        labels += [label.name for label in pr.labels]
```
what is this for? i think we don't read `labels` after in the iteration anymore.

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -192,21 +180,34 @@ def main():
             print(f'no backport label: {pr.number}')
             continue
-        if args.commits and not with_github_keyword_prefix(repo, pr):
+        if not with_github_keyword_prefix(repo, pr) and 'unlabeled' not in args.github_event:
```
why are we permissive here? can we just use 
```py
if not with_github_keyword_prefix(repo, pr) and args.github_event != 'unlabeled':
```
?

tchaikov commented on .github/workflows/add-label-when-promoted.yaml:
```diff
@@ -53,19 +53,28 @@ jobs:
+          echo "Checking labels: $(echo "$labels_json" | jq -r '.[].name')"
+
+          # Check if a valid backport label exists
+          if echo "$labels_json" | jq -e '.[] | select(.name | test("^backport/[0-9]+\\.[0-9]+$"))' > /dev/null; then
```
nit, probably we can use `any()`, just for better readability, as `any()` is exactly what we are trying to test? like:
```shell
if echo "$labels_json" | jq -e 'any(.[] | .name; test("backport/[0-9]+\\.[0-9]+$"))' > /dev/null; then
  

```

see https://jqlang.org/manual/#any

yaronkaikov commented:
v5:
- answered all PR comments

@tchaikov please have another look,

tchaikov commented on .github/workflows/add-label-when-promoted.yaml:
```diff
@@ -53,19 +53,28 @@ jobs:
         env:
           GITHUB_TOKEN: ${{ secrets.AUTO_BACKPORT_TOKEN }}
-        run: python .github/scripts/auto-backport.py --repo ${{ github.repository }} --base-branch ${{ github.ref }} --pull-request ${{ github.event.pull_request.number }} --head-commit ${{ github.event.pull_request.base.sha }}
+        run: python .github/scripts/auto-backport.py --repo ${{ github.repository }} --base-branch ${{ github.ref }} --pull-request ${{ github.event.pull_request.number }} --head-commit ${{ github.event.pull_request.base.sha }} --github-event ${{ github.event.action }}
```
IIUC, if one removes `scylladbbot/backport_error` label, this step should be executed. but the name of it indicates: "Run auto-backport.py when label was added". this is misleading.

tchaikov commented on .github/workflows/add-label-when-promoted.yaml:
```diff
@@ -53,19 +53,28 @@ jobs:
-            echo "backport_label=false" >> $GITHUB_OUTPUT
+            echo "No matching backport label found."
           fi
+          echo "backport_label=false" >> "$GITHUB_OUTPUT"
```
could you please reconsider the suggestion of

> would be better to have a more readable variable name for backport_error_label. i'd suggest use something like "ready_for_backport".

at https://github.com/scylladb/scylladb/pull/23259#discussion_r1995057392 ? as a boolean "backport_label" is confusing in comparison with "ready_for_backport".

tchaikov commented on .github/scripts/auto-backport.py:
```diff
@@ -123,7 +114,7 @@ def backport(repo, pr, version, commits, backport_base_branch):
             repo_local.git.push(fork_repo, new_branch_name, force=True)
             create_pull_request(repo, new_branch_name, backport_base_branch, pr, backport_pr_title, commits,
-                                is_draft=is_draft)
+                                is_draft=is_draft, is_collaborator=is_collaborator)
```
nit, the kwargs style is not necessary here. as the names of the arguments are self-explanatory.

tchaikov reviewed (approved):
lgtm

yaronkaikov commented:
@scylladb/scylla-maint can you please merge this?
